
```{r setup, include=FALSE}
suppressPackageStartupMessages({
    require(BiocStyle)
})
```

# iSEE: Interactive visualization of `SummarizedExperiment` objects

Authors:
    Kevin Rue-Albrecht^[Kennedy Institute of Rheumatology, University of Oxford,
    Headington, Oxford OX3 7FY, UK.],
    Charlotte Soneson^[Friedrich Miescher Institute for Biomedical Research and SIB Swiss Institute of Bioinformatics],
    Federico Marini^[Another Institution],
    Aaron TL Lun^[Another Institution].
    <br/>
Last modified: 4 April, 2019.

## Overview

### Description

This workshop demonstrates the use of the `r Biocpkg("iSEE")` package to create and configure interactive applications for the exploration of various types of genomics data sets (e.g., bulk and single-cell RNA-seq, CyTOF, gene expression microarray).

This workshop will be presented as a lab session that combines an instructor-led live demo, followed by hands-on experimentation guided by completely worked examples and stand-alone notes that participants may continue to use after the workshop.

The instructor-led live demo comprises three parts:

1. Brief lecture on the package concept and functionality
2. Overview of the graphical user interface
3. Instructions to preconfigure `r Biocpkg("iSEE")` apps

The hands-on lab comprises three part:

1. Inspection of single-cell RNA-seq data at various steps of a typical computational workflow, including quality control and dimensionality reduction
2. Addition of custom panels to the user interface for advanced visualization.
3. Additional questions from the participants, including individual use cases and suggestions for future developments

Participants are encouraged to ask questions at any time during the workshop.

### Pre-requisites

* Basic knowledge of [_R_](https://www.r-project.org/about.html) syntax and the use of data-frames
* Familiarity with the `r Biocpkg("SummarizedExperiment")` and `r Biocpkg("SingleCellExperiment")` classes
* Familiarity with the `r CRANpkg("shiny")` CRAN package
* Familiarity with the `r Biocpkg("TENxPBMCData")` package and vignette

Additional background reading about the programming environment, relevant packages, and example use cases:

* Shiny from RStudio: https://shiny.rstudio.com
* SummarizedExperiment paper: https://www.nature.com/articles/nmeth.3252 (Figure 2)
* iSEE paper: https://f1000research.com/articles/7-741/v1

### Participation

Students will participate by following along an [_R markdown_](https://rmarkdown.rstudio.com/) document, and asking questions throughout the workshop.
There is also scope for participants to apply `r Biocpkg("iSEE")` to their own data sets, and fuel the discussion with more questions about specific use cases.

### _R_ / _Bioconductor_ packages used

1. _iSEE_ : https://bioconductor.org/packages/iSEE
2. _SummarizedExperiment_: https://bioconductor.org/packages/SummarizedExperiment
3. _SingleCellExperiment_: https://bioconductor.org/packages/SingleCellExperiment
4. _scater_: https://bioconductor.org/packages/scater
5. _scran_: https://bioconductor.org/packages/scran
6. _TENxPBMCData_: https://bioconductor.org/packages/TENxPBMCData

### Time outline

| Activity                                      | Time |
|-----------------------------------------------|------|
| **Lecture**: Overview of package and concepts | 15m  |
| **Live demo**: the user interface             | 15m  |
| **Lab**: Configuring the app interface        | 20m  |
| **Lab**: A single-cell RNA-seq workflow       | 20m  |
| **Lab**: Custom panels                        | 20m  |
| Additional questions                          | 15m  |

**Total:** 1h45

### Workshop goals and objectives

#### Learning goals

* Recognize the benefits of integrative data containers such as [`SummarizedExperiment`](http://bioconductor.org/packages/SummarizedExperiment) and [`SingleCellExperiment`](http://bioconductor.org/packages/SingleCellExperiment) for downstream analyses and visualization
* Outline the unique features of [_iSEE_](http://bioconductor.org/packages/iSEE/) built upon the [RStudio Shiny](https://shiny.rstudio.com) framework
* Identify biological data that may be combined into insightful graphical outputs
* Utilize interactive GUI components and layouts to efficiently extract information from biological data sets
* Describe how to construct interactive apps and custom panels

#### Learning objectives

* Memorize the key information available in [`SummarizedExperiment`](http://bioconductor.org/packages/SummarizedExperiment) and [`SingleCellExperiment`](http://bioconductor.org/packages/SingleCellExperiment) objects
* Set up a local environment for running [_iSEE_](https://bioconductor.org/packages/iSEE) apps
* Interact with components of the [_iSEE_](https://bioconductor.org/packages/iSEE) user interface to visually inspect and discuss various data sets
* Identify and locate configurable aspects of [_iSEE_](https://bioconductor.org/packages/iSEE) apps
* Practice interactive visualization over a single-cell RNA-sequencing workflow
* Design custom [_iSEE_](https://bioconductor.org/packages/iSEE) panels for advanced use cases
* Imagine use cases and future developments for interactive visualization as part of computational workflows

## Workshop

### Preparation of example scRNAseq data

<!-- What data representation do we use? Sparse matrices, HDF5? -->
<!-- Quick overview of SummarizedExperiment/SingleCellExperiment/other extensions -->
<!-- Compute a range of dimensionality reduction results (perhaps showcase BiocParallel in the process) -->
<!-- Reminder that many types of data can be read directly into SummarizedExperiment objects (with DropletUtils, tximeta, etc) -->

We use the `r Biocpkg("TENxPBMCData")` package to load example data.


```{r, message=FALSE}
library(TENxPBMCData)
```

The man page for the `TENxPBMCData()` function gives an idea of the datasets available.

```{r, eval=FALSE}
help(TENxPBMCData)
```

Load the `"pbmc3k"` dataset.
The first time, this will download the dataset to a local cache, which takes some time, depending on the speed of your internet connection.
Subsequent times, this will load the dataset directly from the local cache.

```{r, message=FALSE, warning=FALSE}
sce <- TENxPBMCData(dataset = "pbmc3k")
```

At this point we can inspect the dataset.

```{r}
sce
```

In particular, this summary view indicates that the following pieces of information are available:

- An assay matrix named `"counts"`
- Row names (i.e. genes) are Ensembl gene identifiers
- Row metadata include the official gene symbol, and the gene symbol used by the 10x CellRanger pipeline
- Column names (i.e., cells) are not initialized and left to `NULL`
- Column metadata include, among others, the cell barcode (`"Barcode"`) and the donor identifier (`"Individual"`).

This is already sufficient information to launch an app and visualize the available data and metadata.

```{r, message=FALSE}
library(iSEE)
if (interactive()){
    iSEE(sce)
}
```

Note that the `iSEE()` function automatically detects the types of information available in the `sce` object, and enable only panels for which the relevant data is available.

### Overview of the _iSEE_ package
<!-- Talk a bit about the usefulness of visualization. -->

### The user interface
<!-- Mention all panel types -->
<!-- Show how to change what is plotted, faceting, changing appearance/color -->
<!-- Transmit selections -->
<!-- Subsampling --> 

### Configuration of the app interface
<!-- Here we could mention the cytometry data, maybe also the 'gating' mode -->
<!-- Interactive modification of panel order/size -->

### Custom panels
<!-- Here we could mention the custom coverage plot, and link to bulk RNA-seq data -->
<!-- Multiple selections -> DE -->

### Reproducibility
<!-- Code export -->

### Sharing and communication
<!-- Tours -->

### Resources
* Bioconductor landing page: [https://bioconductor.org/packages/iSEE/](https://bioconductor.org/packages/iSEE/)
* Publication (F1000Research, 2018): [https://f1000research.com/articles/7-741/v1](https://f1000research.com/articles/7-741/v1)
* Deployed examples: [https://marionilab.cruk.cam.ac.uk/](https://marionilab.cruk.cam.ac.uk/), code at [https://github.com/ltla/isee2018](https://github.com/ltla/isee2018)
* Custom panel examples: [https://github.com/kevinrue/iSEE_custom](https://github.com/kevinrue/iSEE_custom)
* Development version (bug reports etc): [https://github.com/csoneson/isee](https://github.com/csoneson/isee)
